websockets

bgwdotdev opened the pull request:
Here's a first pass, scrappy implementation of mists websockets into wisp. 

Still figuring out if there's a better way to do this and discussing with rawhat a little on approaches.

This currently works though. Feedback on approach most welcome.

Option was only added due to the test stuff, not sure how do deal with that yet but maybe that's solving the wrong problem.

lpil commented on src/wisp.gleam:
```diff
@@ -1832,5 +1848,27 @@ pub fn create_canned_connection(
   )
 }
+
+pub fn websocket(
+  req: Request,
+  handler handler: fn(a, mist.WebsocketConnection, mist.WebsocketMessage(b)) ->
+    actor.Next(b, a),
+  on_init on_init: fn(mist.WebsocketConnection) ->
+    #(a, Option(process.Selector(b))),
+  on_close on_close: fn(a) -> Nil,
+) -> Response {
```
We need to not leak any Mist types into the application stack. Wisp doesn't depend on Mist beyond that convenience function to convert a Wisp handler to a Wisp one.

lpil commented on src/wisp.gleam:
```diff
@@ -145,6 +153,8 @@ pub type Body {
   ///
   Empty
+
+  Websocket(process.Selector(process.ProcessDown))
```
What is this? It's undocumented.

lpil commented on src/wisp.gleam:
```diff
@@ -605,10 +615,15 @@ pub opaque type Connection {
     read_chunk_size: Int,
     secret_key_base: String,
     temporary_directory: String,
+    mist_conn: Option(mist.Connection),
```
What is this for, and why is it optional?

This also makes the connection depend on Mist specifically. It'll need to be redesigned to not do so.

bgwdotdev replied:
This is the real challenging one as I'm not sure we can actually get rid of this without mist changing the  public API around websockets. Specifically mist depends on these fields[1] of the connection which we can maybe extract but I'm not sure we can satisfy the connection type without storing it, and as it is a 'potentially' unstable internal type, that may be a risky maneuver and I'm not sure your opinion on that?

Perhaps I need to have a deeper discussion with rawhat about other paths.
 
1: https://github.com/rawhat/mist/blob/master/src/mist/internal/http.gleam#L38-L39

lpil reviewed (commented):
Thank you! This couples Wisp very tightly to Mist, which it carefully does not do so far. We'll need to decouple it, I've left some notes inline.

bgwdotdev commented:
I've now created a pure wisp public api for this which mostly wraps the mist types. 

Docs (and likely type names/api) still need work but wanted to verify if the approach is more suitable for wisp before attempting to polish things. 

Hopefully the Socket type would allow wrapping various connection types that may be required by potentially different backends as appropriate?

Let me know if I'm on the right track or missing a beat.

Thanks :)

lpil commented on src/wisp.gleam:
```diff
@@ -1832,5 +1848,103 @@ pub fn create_canned_connection(
+  WSBinary(BitArray)
+  WSClosed
+  WSShutdown
+  WSCustom(a)
```
nit: Gleam's convention says these should be `WsCustom` etc

lpil commented on src/wisp.gleam:
```diff
@@ -1832,5 +1848,103 @@ pub fn create_canned_connection(
+}
+
+pub opaque type WebsocketConnection {
+  WebsocketConnection(mist.WebsocketConnection)
```
This depends on Mist here, but there must be no Mist specific code past the `mist_service` function.

lpil commented on src/wisp.gleam:
```diff
@@ -1832,5 +1848,103 @@ pub fn create_canned_connection(
+    actor.Next(b, a),
+  on_init on_init: fn(WebsocketConnection) -> #(a, Option(process.Selector(b))),
+  on_close on_close: fn(a) -> Nil,
+) -> Response {
```
Same here, there must be no Mist specific code by this point.

lpil reviewed (commented):
Nice!! I like how this is coming along.

There's still some dependency of Mist that needs to be removed. Soon the Mist specific code will be moved to another package, so there can be no dependency here.

Another problem to solve is how do we make it so a Wisp application can only upgrade to a websocket connection if the webserver being used supports it? For example, Elli and CGI don't support websockets, so the programmer must not have the ability to construct a websocket response in those cases. Perhaps it could be that there's some secondary capability value that gets passed in by websocket capable servers, and that is need to upgrade within the application. Whatever the solution the same pattern will need to be usable for server sent events too.

Another thing is how do we write tests for websocket using applications too.

Is there any other problems to solve or other considerations?

bgwdotdev commented:
Okay. Had a big stab at moving all the mist stuff into a separate module.

I've fully removed the import mist from the main wisp.gleam file.

Websockets are also now implemented with a reference example in the test file.

I've had to make a couple things in the wisp.gleam file public that I'm not sure we want to keep that way (mostly around `make_connection`).

Everything needs a tidy, documentation and the test cases fleshed out but hopefully it's looking more like how we want it to be ^^?

Definitely not finished working on it but keen to take any feedback generally at this stage!

lpil commented on gleam.toml:
```diff
@@ -9,6 +9,10 @@ links = [
 
+internal_modules = [
+  "wisp/internal",
+]
```
This is a subset of the default value so you can remove this

bgwdotdev commented:
Okay! I think I'm feeling happy with the general api design and split of things now and was thinking of moving on to docs.

I was thinking of creating a websocket error type within wisp as well, though looking at the errors provided from mist via glisten, I'm thinking we might want to just cover important ones like `Closed` and `Timeout` maybe and then the others can fall under some more generic socket error that left for the user to handle? (here's the variants for reference https://hexdocs.pm/glisten/3.0.0/glisten/socket.html#SocketReason)

bgwdotdev commented:
left to do:

update examples
write example using websocket

done?

lpil commented on gleam.toml:
```diff
@@ -20,6 +20,8 @@ mist = ">= 0.13.0 and < 2.0.0"
 marceau = ">= 1.1.0 and < 2.0.0"
 logging = ">= 1.0.0 and < 2.0.0"
+gleam_otp = ">= 0.10.0 and < 1.0.0"
+stratus = ">= 0.9.0 and < 1.0.0"
```
This should not be a dependency. Could be a dev one

lpil commented on src/wisp.gleam:
```diff
@@ -47,68 +49,9 @@ import simplifile
-  }
+@deprecated("use wisp_mist.handler instead")
+pub fn mist_handler() {
+  todo as "document move"
```
Delete this please 🙏

lpil commented on src/wisp/internal.gleam:
```diff
@@ -0,0 +1,84 @@
+import gleam/bit_array
```
Delete this module please

lpil commented on src/wisp/wisp_mist.gleam:
```diff
@@ -0,0 +1,176 @@
+///     |> mist.new
+///     |> mist.port(8000)
+///     |> mist.start_http
+///   process.sleep_forever()
```
This documentation doesn't match what the function does now

lpil commented on test/mist_test.gleam:
```diff
@@ -0,0 +1,111 @@
+import wisp/wisp_mist
+
+pub fn websocket_test() {
+  process.start(fn() { webserver() }, True)
```
You can call `mist.start_http` and remove the extra layer of process wrapping here

lpil commented on test/mist_test.gleam:
```diff
@@ -0,0 +1,111 @@
+    #(0, None)
+  }
+  let handler = fn(state, conn, msg) {
+    io.debug(msg)
```
There's 5 debug statements left in here

lpil commented on src/wisp/wisp_mist.gleam:
```diff
@@ -0,0 +1,176 @@
+  }
+}
+
+/// Sends data to a websocket connection from a wisp websocket send spec
```
What does "from a Wisp websocket send spec" mean?

Capital letter for Wisp, and full stops at the end of sentences please.

lpil commented on src/wisp.gleam:
```diff
@@ -1829,8 +1727,146 @@ pub fn create_canned_connection(
+  WsBinary(BitArray)
+  WsClosed
+  WsShutdown
+  WsCustom(a)
```
Document these please 🙏

lpil commented on src/wisp.gleam:
```diff
@@ -1829,8 +1727,146 @@ pub fn create_canned_connection(
+// Websockets
+//
+
+/// The messages possible to receive to and from a websocket handler.
```
Don't these only come from the client?

bgwdotdev replied:
No, the `WsCustom` type is called when messages are sent to the optional subject/selector created during the websocket handler `on_init`. This is how you generally send messages from within your application to a connected websocket. 

For example, your server might have a `broadcast_clients` function that sends a message to all websocket handlers subject, which it will receive on WsCustom, and then maybe you forward that message onto the connected websocket(s) themselves! If that makes sense?

lpil replied:
Sorry, I mean this says the handler sends these, but it receives them

bgwdotdev replied:
Oh right, I see what you mean, of course! Corrected this now :)

lpil commented on src/wisp/wisp_mist.gleam:
````diff
@@ -0,0 +1,176 @@
+/// }
+/// ```
+pub fn handler(
+  handler: fn(wisp.Request, wisp.Ws(mist.Connection)) -> wisp.Response,
````
Mist specific types leak into the application here. A Wisp websocket application should be able to be run on any web socket capable web server without modification, but in this case the type would need to be changed.

bgwdotdev replied:
I can wrap the `mist.Connection` in a `wisp_mist.Connection` type to hide the mist implementation but I don't think it's possible to remove this altogether as this holds the http socket information which, when a request is made to a websocket, it uses to upgrade the socket to a websocket and then hold onto it to keep the connection alive (to my understanding).

This socket connection implementation will likely be unique per websocket server's approach to managing the connection.

Unless I'm missing something, there's no way to hide this without at least wisp/internal depending on mist/whatever-other-server.

My thought was if you need to swap wisp_mist for whatever, this would involve changing. 

```gleam
wisp.Ws(wisp_mist.Connection) => wisp.Ws(other_server.Connection)
```

or maybe if they were imported with an alias, it wouldn't change the api :?

```gleam
import wisp_mist as server => import other_server as server

wisp.Ws(server.Connection)
```

lpil replied:
This is still specific server implementations leaking into Wisp core, making Wisp applications no longer web server agnostic. It's a hard requirement for the same application to be deployable on different web servers with the same capabilities.

bgwdotdev replied:
Latest changes now stop these types from leaking but I can't find a way to do it without the handler function at least returning a function which takes in the required functions to start a websocket actor:  `fn(WebsocketHandler(a,b) -> wisp.Response` 

The problem with this design is you can then only have one solid type for your applications websockets that you need to specific in your handler function.

Spent a lot of time on this and can't seem to find a way around it. Might just need to sleep on it but if you have any ideas, it'd be very welcome ^^. 

Any other option I think of just seem to lead to mist leaks :<

bgwdotdev replied:
outside of using `coerce`, which works but obvious is theoretically unsafe

lpil commented on src/wisp.gleam:
```diff
@@ -1829,8 +1727,146 @@ pub fn create_canned_connection(
+type WebsocketConnection(c) =
+  internal.WebsocketConnection(c)
+
+/// For web socket capable servers to connect to clients
```
I don't understand what this is from the documentation, could it be expanded please 🙏 

The name isn't very clear either. It doesn't appear to be a websocket?

bgwdotdev replied:
hopefully makes more sense now ^^?

lpil commented on test/mist_test.gleam:
```diff
@@ -0,0 +1,111 @@
+}
+
+fn ws_handler(req: wisp.Request, ctx: Context) {
+  let on_init = fn(conn: wisp.WebsocketConnection(mist.WebsocketConnection)) {
```
I think the Mist specific types leak here too?

lpil reviewed (commented):
Nice! I've left a handful of notes inline

bgwdotdev commented:
Thanks for the feedback+review, been busy with work so not been able to progress it, off next week though so will try push this along to finish line and address all your notes :)! Much appreciated.

bgwdotdev commented:
the mist module split only has now been created as another pr ahead of the work here if we want to tackle that bit first :)

bgwdotdev commented:
just as a note, rebase off main now so diff should be easier

lpil commented on examples/12-websocket-chatroom/src/app/websocket.gleam:
```diff
@@ -0,0 +1,92 @@
+import app/chatroom
```
Remove this example please. It works but it's not the best way to implement this and has some failure cases which can leak memory or otherwise go wrong, and I wouldn't want people to copy it rather than studying OTP to figure out a more reliable way.

bgwdotdev replied:
Good point, makes sense, deleted this. Perhaps worth us revisiting at a later date to make a quality example as seems lots of folks look for a reference on this type of thing :)

lpil commented on src/wisp.gleam:
```diff
@@ -60,6 +62,10 @@ pub type Body {
+  /// Upgrades the socket to a websocket, is only used internally by the
+  /// web server and should not be called directly.
+  ///
+  Websocket(process.Selector(process.ProcessDown))
```
Why does it contain `process.Selector(process.ProcessDown)`?

bgwdotdev replied:
Honestly, I'm not sure -- I don't really understand the underlying otp stuff going on here ^^; I'll try to look into it to understand it more but all I know is that if you remove it, the websocket just automatically closes after the first message

lpil replied:
What do you mean? Types don't change runtime behaviour like that.

bgwdotdev replied:
Hm, not sure, could be bad memory on my part, I'll give it another test and see :)

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,167 @@ pub fn create_canned_connection(
+/// function. It is required to turn a http connection into an
+/// active websocket (`WebsocketConnection`).
+///
+pub type WsCap(state, msg) {
```
Capability please

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,167 @@ pub fn create_canned_connection(
+/// This connection is used from within a websocket handler function to
+/// send data to the client via `SendText` or `SendBinary`
+///
+pub type WsConn =
```
Connection please, we don't abbreviate in Gleam.

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,167 @@ pub fn create_canned_connection(
+/// This type will need to be passed to your webserver of choice websocket
+/// function, such as `wisp_mist.websocket`.
+///
+pub type WebsocketHandler(state, msg) {
```
In some places its Ws, some it's Websocket. Which do we want? Normally we don't abbreviate in Gleam so I'm leaning towards Websocket

bgwdotdev replied:
Apologies, I've now standardised this to `WsXyz` format. I'm happy to go either way on this though. I feel `WebsocketXyz` get really cumbersome though that may just be because I've been writing a lot of variations of it while trying to figure out an api hah.

I would entirely defer to your preference on this and happy to change as appropriate. The only other though I has was moving the websocket stuff into a `wisp/websocket.gleam` module and perhaps removing Ws/Websocket entirely from all the names?

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,167 @@ pub fn create_canned_connection(
+/// }
+///
+/// fn websocket(req, ws, server) -> Response {
+///   wisp.WebsocketHandler(req, ws, handler, on_init(_, server), on_close)
```
This has 5 parameters but the code has 3?

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,167 @@ pub fn create_canned_connection(
+/// active websocket (`WebsocketConnection`).
+///
+pub type WsCap(state, msg) {
+  WsCap(fn(Request, WebsocketHandler(state, msg)) -> Response)
```
What's this inner value for? I couldn't see it used anywhere

bgwdotdev replied:
Apologies, the  examples had fallen out of line with the new api, it was present in the test but I've actually added a new `wisp.websocket` function which I think better shows how this is used. I've also updated all the test/examples to use this

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,167 @@ pub fn create_canned_connection(
+///     }
+///   }
+///   let on_close = fn(_state) { Nil }
+///   wisp.WebsocketHandler(req, ws, handler, on_init, on_close)
```
This doesn't match what it does in the code

lpil reviewed (commented):
Thank you! I've left a series of questions

bgwdotdev commented:
Ok, I think I've updated everything! docs, examples, tests hopefully all covered :)

Assuming there was no further changes required, the only issue left I think would be how we want to do errors. I'm not happy with the error handling I've added but also not sure we want to copy all the errors available in the mist implementation. If you have an idea on how you'd prefer this was present to the user, happy to follow that :)

otherwise if just translating over all the glisten socket errors makes sense then happy to do that as well:
https://hexdocs.pm/glisten/3.0.0/glisten/socket.html#SocketReason

giovannibonetti-work commented:
I suppose @lpil is very busy, but seeing he didn't even react with an emoji yet, I think he might have missed the notification.
Or maybe he is on vacation, I don't know. Anyway, I'm just trying to be helpful by bumping up this thread and mentioning him.

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,188 @@ pub fn create_canned_connection(
+/// send data to the client via `WsSendText` or `WsSendBinary`
+///
+type WsConnection =
+  fn(WsSend) -> Result(Nil, WsError)
```
Why is a connection a function?

bgwdotdev replied:
This was changed into a callback to avoid leaking the mist websocket connection type or creating a mist specific implementation of `send` due to the types required.

To avoid this it is we return a callback from our mist on_init wrapper that has the connection hidden away inside it, so you just pass your data to the WsConnection type and it takes care of attaching the mist (or otherwise) connection type, on your behalf.

Maybe another name would make more sense?  I'm not able to see another api design to get around this problem if we don't want it to be a function unfortunately

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,188 @@ pub fn create_canned_connection(
+/// function. It is required to turn a http connection into an
+/// active websocket (`WsConnection`).
+///
+pub type WsCapability(state, msg) {
```
Should the internals of this be public? How would this dissuade users from constructing it directly?

Why does the capability have type parameters and contain the handler? Do you want to restrict the websocket processes to have specific messages and state for other implementations?

bgwdotdev replied:
> Should the internals of this be public? 

Due to a circular import issue, I was unable to put this into internal (as then internal would depend on request and response from wisp.gleam and wisp.gleam would depend on internal.gleam to export it as an opaque type) but it is required to be construct-able by mist_wisp or other server implementations -- I'll look into seeing if I can create another internal module and have it act as a middle man or diamond configuration or whatever the correct term for that is ^^;

>  How would this dissuade users from constructing it directly?

In the current implementation, to construct it you need to import `internal` which I was hoping would be enough to dissuade folks.

> Why does the capability have type parameters and contain the handler? Do you want to restrict the websocket processes to have specific messages and state for other implementations?

This ending up being an unfortunate side-effect of being the only way I could find to remove any mist types leaking due to the need to store the mist connection info for the socket. This way we can essentially hide the mapping of `wisp.Request -> mist.Connection -> wisp.Response` but it means the capability needs to return a callback which takes in the single extra arg of the websocket handler.

I actually don't really want it to work like this at all as it causes awkwardness if you want to use more than one websocket handler in your wisp server at once but I'm unable to see any other approach to dealing with this in the wisp_mist.handler

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,188 @@ pub fn create_canned_connection(
+  WsSendBinary(binary: BitArray)
+}
+
+/// TODO: Placeholder error type, flesh out.
```
Is this done?

bgwdotdev replied:
from previous message for your consideration (as I feel like you wouldn't want to copy all those socket errors but I'm not sure, if you have a quick ye/naa confirmation that works for me)

> the only issue left I think would be how we want to do errors. I'm not happy with the error handling I've added but also not sure we want to copy all the errors available in the mist implementation. If you have an idea on how you'd prefer this was present to the user, happy to follow that :)
>
>  otherwise if just translating over all the glisten socket errors makes sense then happy to do that as well:
https://hexdocs.pm/glisten/3.0.0/glisten/socket.html#SocketReason

lpil commented on src/wisp.gleam:
```diff
@@ -1876,3 +1882,188 @@ pub fn create_canned_connection(
+  WsErrClosed
+  WsErrTimeout
+  WsErrTerminated
+  WsErrOther(String)
```
No abbreviations please. Put Error at the end of the names.

What's Other for? Document these please.

lpil commented on src/wisp/wisp_mist.gleam:
```diff
@@ -22,19 +27,26 @@ import wisp/internal
+///
 pub fn handler(
-  handler: fn(wisp.Request) -> wisp.Response,
+  handler: fn(wisp.Request, wisp.WsCapability(state, msg)) -> wisp.Response,
```
No breaking changes to this API please. We said we'd have a new function here.

bgwdotdev replied:
Oh, apologies, I had it in my backwards in my head for whatever reason, will fix

lpil commented on src/wisp/internal.gleam:
```diff
@@ -84,3 +82,14 @@ pub fn random_string(length: Int) -> String {
+/// A type used purely to block the user constructing a `wisp.WsCapability`
+/// via the public api as it should only be constructed by the web server if it
+/// supports websockets.
+pub type WsCapability {
```
This should be opaque

bgwdotdev replied:
As noted earlier due to issues of circular import and needing to construct this in wisp_mist.gleam it is not opaque, will see if I can shuffle the module around to allow this

lpil commented on src/wisp.gleam:
````diff
@@ -1876,3 +1882,188 @@ pub fn create_canned_connection(
+/// "pong" |> wisp.WsSendText |> conn()
+/// ```
+///
+pub type WsSend {
````
Is there a reason not to always sent bit arrays?

bgwdotdev replied:
No reason I'm aware of, will give it a test and see :)

lpil commented:
Have just been giving this a try. How does one create more than one webhook handler? The type parameter means they have to all have the same types, so I can't create a second one for another route.

bgwdotdev commented:
Yeah, this is exactly the big problem and sticking point I can't find away around in this ( or any I can think of :( ) design without leaking mist types into wisp/user applications.

Essentially, you'd have to make each websockets type present in a single sum type for each the message and state as we're currently forced to lock in the WsCapability type in the handler.

I explored "safely" casting the types under the hood which while it worked and I _believe_ was safe from being able to be misused, as you've previously stated and I do agree with, it does undermine the type system and a solution away from this would be the way to go.

Unfortunately I can't seem to find a way to set this type any later or earlier than the handler without leaking mists connection.  It's probably a skill issue but I've gone over it and over it many times to no avail sadly.
